src_feedback/elf.rs

```diff
@@ -0,0 +1,666 @@
+    let mut buf = vec![0; size];
+
+    file.seek(SeekFrom::Start(off))?;
+    file.read_exact(buf.as_mut_slice())?;
+
+    Ok(buf)
+}
+
+fn read_elf_header(file: &mut File) -> Result<Elf64_Ehdr, Error> {
+    const DSZ: usize = mem::size_of::<Elf64_Ehdr>();
+    let mut buf = Box::new([0_u8; DSZ]);
+
+    let buf_m: &mut [u8; DSZ] = buf.borrow_mut();
+    file.read_exact(buf_m)?;
+
+    let ehdr: Box<Elf64_Ehdr> = unsafe {
+	// A complicated type casting hacking!!
+	let ehdr_raw_ptr = (Box::leak(buf) as *mut u8) as *mut Elf64_Ehdr;
+	Box::from_raw(ehdr_raw_ptr)
+    };
+
+    Ok(*ehdr)
+}
+
+fn read_elf_sections(file: &mut File, ehdr: &Elf64_Ehdr) -> Result<Vec<Elf64_Shdr>, Error> {
+    const HDRSIZE: usize = mem::size_of::<Elf64_Shdr>();
+    let off = ehdr.e_shoff as usize;
+    let num = ehdr.e_shnum as usize;
+
+    let mut buf = read_u8(file, off as u64, num * HDRSIZE)?;
+
+    let shdrs: Vec<Elf64_Shdr> = unsafe {
+	let shdrs_ptr = buf.as_mut_ptr() as *mut Elf64_Shdr;
+	buf.leak();
+	Vec::from_raw_parts(shdrs_ptr, num, num)
+    };
+    Ok(shdrs)
+}
+
+fn read_elf_program_headers(file: &mut File, ehdr: &Elf64_Ehdr) -> Result<Vec<Elf64_Phdr>, Error> {
+    const HDRSIZE: usize = mem::size_of::<Elf64_Phdr>();
+    let off = ehdr.e_phoff as usize;
+    let num = ehdr.e_phnum as usize;
+
+    let mut buf = read_u8(file, off as u64, num * HDRSIZE)?;
+
+    let phdrs: Vec<Elf64_Phdr> = unsafe {
+	let phdrs_ptr = buf.as_mut_ptr() as *mut Elf64_Phdr;
+	buf.leak();
+	Vec::from_raw_parts(phdrs_ptr, num, num)
+    };
+    Ok(phdrs)
+}
+
+fn read_elf_section_raw(file: &mut File, section: &Elf64_Shdr) -> Result<Vec<u8>, Error> {
+    read_u8(file, section.sh_offset as u64, section.sh_size as usize)
+}
+
+fn read_elf_section_seek(file: &mut File, section: &Elf64_Shdr) -> Result<(), Error> {
+    file.seek(SeekFrom::Start(section.sh_offset as u64))?;
+    Ok(())
+}
+
+fn read_elf_section_offset_seek(file: &mut File, section: &Elf64_Shdr, offset: usize) -> Result<(), Error> {
+    if offset as u64 >= section.sh_size {
+	return Err(Error::new(ErrorKind::InvalidInput, "the offset is too big"));
+    }
+    file.seek(SeekFrom::Start(section.sh_offset as u64 + offset as u64))?;
+    Ok(())
+}
+
+fn get_elf_section_name<'a>(sect: &Elf64_Shdr, strtab: &'a [u8]) -> Option<&'a str> {
+    extract_string(strtab, sect.sh_name as usize)
+}
+
+struct Elf64ParserBack {
+    ehdr: Option<Elf64_Ehdr>,
+    shdrs: Option<Vec<Elf64_Shdr>>,
+    shstrtab: Option<Vec<u8>>,
+    phdrs: Option<Vec<Elf64_Phdr>>,
+    symtab: Option<Vec<Elf64_Sym>>, // Sorted symtab
+    symtab_origin: Option<Vec<Elf64_Sym>>, // The copy in the same order as the file
+    strtab: Option<Vec<u8>>,
+}
+
+/// A parser against ELF64 format.
+///
+pub struct Elf64Parser{
+    file: RefCell<File>,
+    backobj: RefCell<Elf64ParserBack>,
+}
+
+impl Elf64Parser {
+    pub fn open_file(file: File) -> Result<Elf64Parser, Error> {
```

Comment:
nit: I think this function name is a little confusing because it doesn't open a file, it just wraps it in a Elf64Parser. Would something like `init_parser` be more clear?

Reply:
agree